fri/src/verifier.rs

```diff
@@ -267,3 +306,137 @@ where
+/// - `params`: The FRI parameters.
+/// - `log_global_max_height`: The log of the maximum height of the input matrices.
+/// - `index`: The index at which to open the functions.
+/// - `input_proof`: A vector of batch openings with each opening containing a
+///   list of opened values for a collection of matrices along with a batched opening proof.
+/// - `alpha`: The challenge used to combine the functions.
+/// - `input_mmcs`: The input multi-matrix commitment scheme.
+/// - `commitments_with_opening_points`: A vector of joint commitments to collections of matrices
+///   and openings of those matrices at a collection of points.
+#[inline]
+fn open_input<Val, Challenge, InputMmcs, FriMmcs>(
+    params: &FriParameters<FriMmcs>,
+    log_global_max_height: usize,
+    index: usize,
+    input_proof: &[BatchOpening<Val, InputMmcs>],
+    alpha: Challenge,
+    input_mmcs: &InputMmcs,
+    commitments_with_opening_points: &[CommitmentWithOpeningPoints<
+        Challenge,
+        InputMmcs::Commitment,
+        TwoAdicMultiplicativeCoset<Val>,
+    >],
+) -> Result<FriOpenings<Challenge>, FriError<FriMmcs::Error, InputMmcs::Error>>
+where
+    Val: TwoAdicField,
+    Challenge: ExtensionField<Val>,
+    InputMmcs: Mmcs<Val>,
+    FriMmcs: Mmcs<Challenge>,
+{
+    // For each log_height, we store the alpha power and compute the reduced opening.
+    // log_height -> (alpha_pow, reduced_opening)
+    let mut reduced_openings = BTreeMap::<usize, (Challenge, Challenge)>::new();
+
+    // For each batch commitment and opening proof
+    for (batch_opening, (batch_commit, mats)) in zip_eq(
+        input_proof,
+        commitments_with_opening_points,
+        FriError::InvalidProofShape,
+    )? {
+        // Find the height of each matrix in the batch.
+        // Currently we only check domain.size() as the shift is
+        // assumed to always be Val::GENERATOR.
+        let batch_heights = mats
+            .iter()
+            .map(|(domain, _)| domain.size() << params.log_blowup)
+            .collect_vec();
+        let batch_dims = batch_heights
+            .iter()
+            // TODO: MMCS doesn't really need width; we put 0 for now.
+            .map(|&height| Dimensions { width: 0, height })
+            .collect_vec();
+
+        if let Some(batch_max_height) = batch_heights.iter().max() {
+            let log_batch_max_height = log2_strict_usize(*batch_max_height);
+            let bits_reduced = log_global_max_height - log_batch_max_height;
+            let reduced_index = index >> bits_reduced;
+
+            // Verify that the opened values match the commitment.
+            input_mmcs.verify_batch(
+                batch_commit,
+                &batch_dims,
+                reduced_index,
+                batch_opening.into(),
+            )
+        } else {
+            // Empty batch?
+            input_mmcs.verify_batch(batch_commit, &[], 0, batch_opening.into())
+        }
+        .map_err(FriError::InputError)?;
```

Comment:
Can't we replace by something like this?

```suggestion
        let reduced_index = batch_heights
            .clone()
            .max()
            .map(|h| index >> (log_global_max_height - log2_strict_usize(h)))
            .unwrap_or(0);

        input_mmcs
            .verify_batch(batch_commit, &batch_dims, reduced_index, batch_opening.into())
            .map_err(FriError::InputError)?;
```